aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h

```diff
@@ -241,52 +272,75 @@ void _validate_compressed_sparse_indices_kernel(
-              index_t batch_idx) -> index_t {
-              // Invariant 5.1
-              _check_first_cidx_is_zero<cdim_name, index_t>(cidx_first, zero);
-              // Invariant 5.2
-              _check_last_cidx_is_nnz<cdim_name, index_t>(cidx_last, nnz);
-              // Invariant 5.3
-              _check_cidx_nondecreasing_locally_bounded_sequence<cdim_name, index_t>(cidx_curr, cidx_next, zero, dim);
-              // Invariant 5.6
-              // NOTE: the implementation below is sync-less, but, unfortunately,
-              // work is not guaranteed to be well-balanced between different threads.
-              // idx is contiguous and of shape (..., nnz), so batches are multiples of nnz apart.
-              const auto* RESTRICT ptr_idx_batch = ptr_idx + batch_idx * nnz;
-              _check_idx_sorted_distinct_vals_slices_with_cidx<cdim_name, index_t>(
-                  ptr_idx_batch, cidx_curr, cidx_next);
-              return 0;
-            }
-        );
-    });
+                    .set_check_mem_overlap(false)
+                    .add_owned_output(dummy.expand_as(cidx_curr))
+                    .add_input(cidx_first)
+                    .add_input(cidx_last)
+                    .add_input(cidx_curr)
+                    .add_input(cidx_next)
+                    .add_input(batch_idx)
+                    .build();
+
+    AT_DISPATCH_INDEX_TYPES(
+        idx.scalar_type(),
+        NAME,
+        [&iter, &idx, dim, nnz, idx_ndims, &idx_sizes, &idx_strides]() {
+          const auto* RESTRICT ptr_idx = idx.data_ptr<index_t>();
+          const auto zero = index_t{0};
+          KernelLauncher::launch(
+              iter,
+              [zero, dim, nnz, idx_ndims, idx_sizes, idx_strides, ptr_idx] FUNCAPI(
+                  index_t cidx_first,
+                  index_t cidx_last,
+                  index_t cidx_curr,
+                  index_t cidx_next,
+                  index_t batch_idx) -> index_t {
+                // Invariant 5.1
+                _check_first_cidx_is_zero<cdim_name, index_t>(cidx_first, zero);
+                // Invariant 5.2
+                _check_last_cidx_is_nnz<cdim_name, index_t>(cidx_last, nnz);
+                // Invariant 5.3
+                _check_cidx_nondecreasing_locally_bounded_sequence<
+                    cdim_name,
+                    index_t>(cidx_curr, cidx_next, zero, dim);
+                // Invariant 5.6
+                // NOTE: the implementation below is sync-less, but,
+                // unfortunately, work is not guaranteed to be well-balanced
+                // between different threads.
+                int64_t idx_offset = 0;
+                // assuming idx contiguity per batch:
+                int64_t tmp = batch_idx * idx_sizes[idx_ndims - 1];
+                for (int i = idx_ndims - 1; i >= 0; i--) {
+                  idx_offset += (tmp % idx_sizes[i]) * idx_strides[i];
+                  tmp /= idx_sizes[i];
+                }
```

Comment:
Division and mod are slow on CUDA, especially compared to all other ops in the kernel. I am fine with that for now, but it would not hurt to optimize this case for contiguous inputs.

Reply:
I can eliminate `mod` here but overall this feels like an over-optimization considering that the invariants validation can be disabled if maximum performance is required.

Reply:
Eliminated `mod`.